proposals/0000-reconsider-typealias-in-protocol.md

```diff
@@ -0,0 +1,138 @@
+
+**Swift 4.1** began preparations to eliminate the feature by adding certain warnings and conditionally disallowing such type aliases in protocol extensions by raising a `type look up ambiguity` error towards appliances of the conflicting type name, if any.
+
+#### `#1`
+```swift
+protocol P {associatedtype A}
+protocol P1: P {typealias A = Int}
+// warning: typealias overriding associated type 'R' from protocol 'P' is better expressed as same-type constraint on the protocol
+```
+#### `#2`
+```swift
+protocol P {typealias A = Int}
+protocol P1: P {associatedtype A}
+// warning: associated type 'R' is redundant with type 'R' declared in inherited protocol 'P'
+```
+#### `#3`
+```swift
+protocol P {
+  associatedtype A
+  func foo() -> A // error: 'A' is ambiguous for type lookup in this context
+}
+extension P {typealias A = Int}
+```
+### Solution
+Now that we have powerful and expressive where clauses for protocol declarations,
+I propose to bring consistency to the usage of type aliases within protocols and protocol extensions.
+
+* [Type aliases in protocol extensions](#type-aliases-in-protocol-extensions). There are two options:
+    * Permit type aliases to act as default values for associated types in protocol extensions. This is the preferred           solution. 
+    * Change the error message to propose the current syntax instead of simply producing **#3**,
+    which is actually a generic error.
+* Type aliases in protocols should be always discouraged in **Swift 4.2** and always disallowed in **Swift 5**, the same way we disallow any implementation details.
+* Warning **#1** should be amended to a generic `typealias should/must not be declared within a protocol` warning/error for **Swift 4.2** and **Swift 5** respectively with fix-it options to either switch to an `associatedtype` or impose a same-type constraint on the protocol.
+* Warning **#2** should be superseded by the `typealias should/must not be declared within a protocol` warning/error in **Swift 4.2** and **Swift 5** respectively.
+
+## Detailed design
+
+#### Non-conflicting type alias in protocol
+```swift
+protocol P {
+ typealias A = Int // fixit: typealias -> associatedtype
+ /// Swift 4.2
+ /// warning: type alias should not be declared within a protocol;
+ /// move to an extension or use 'associatedtype' to define an associated type requirement 
+ /// Swift 5
+ /// error: type alias must not be declared within a protocol;
+ /// move to an extension or use 'associatedtype' to define an associated type requirement
+}
+```
+#### Type alias conflicting with inherited associated type
+
+```swift
+protocol P {associatedtype A}
+protocol P1: P {    // fix-it: where A == Int
+  typealias A = Int // fix-it: typealias -> associatedtype
+  /// Swift 5
+  /// error: typealias must not be declared within a protocol
+  /// note: use 'associatedtype' to specify a default value for 'A' from protocol 'P'
+  /// note: use a same-type constraint on the protocol
+}
+```
```

Comment:
It's worth talking about what happens if the `typealias` is in a constrained protocol extension, e.g.,

```swift
protocol P {
  associatedtype A
  associatedtype A2
}

protocol Q {
  associatedtype B
}

extension P where Self.A2: Q {
  typealias A = A2.B
}
```